nomt/src/beatree/iterator.rs

```diff
@@ -0,0 +1,543 @@
+//! Database iterators over the Beatree.
+
+// TODO: remove this once used.
+#![allow(unused)]
+
+use std::{
+    cmp::Ordering,
+    ops::{Range, RangeFrom},
+    sync::Arc,
+};
+
+use imbl::ordmap::{Iter as OrdMapIter, OrdMap};
+use nomt_core::trie::ValueHash;
+
+use super::{
+    allocator::PageNumber,
+    branch::node::{get_key, BranchNode},
+    index::Index,
+    leaf::node::LeafNode,
+    Key, ReadTransaction, ValueChange,
+};
+
+/// An iterator over the state of the beatree at some particular point.
+///
+/// This combines the in-memory overlays with the state of the leaf pages on the disk.
+/// This iterator does not handle the fetching of pages internally, but instead provides the needed
+/// page numbers. The [`ReadTransaction`]` will provide facilities for dispatching I/O with a
+/// provided handle.
+///
+/// This is not a normal Rust iterator, due to its need to block. Furthermore, it is a streaming
+/// iterator which does not clone or copy its outputs, rather returning them as borrowed. This means
+/// that the standard iterator combinators can't be used with it, making it less versatile than a
+/// typical Rust iterator.
+pub struct BeatreeIterator {
+    memory_values: StagingIterator,
+    leaf_values: LeafIterator,
+}
+
+impl BeatreeIterator {
+    pub(super) fn new(tx: &ReadTransaction, start: Key, end: Option<Key>) -> Self {
+        BeatreeIterator {
+            memory_values: StagingIterator::new(tx, start, end),
+            leaf_values: LeafIterator::new(tx.bbn_index.clone(), start, end),
+        }
+    }
+
+    /// Get the next value from the iterator.
+    ///
+    /// This may either return a value or a marker indicating that it is blocked on a page load.
+    /// The page to load can be determined with [`BeatreeIterator::next_pns`].
```

Comment:
Probably a non-updated comment, should be something like 'the next pages to be loaded are provided by `needed_page`'